File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> brackets;

    for (int i = 0; i < s.length(); i++)
    {
        char c = s[i];

        if (c == '(' || c == '{' || c == '[')
        {
            brackets.push_back(c);
        }
        
        if (brackets.size() <= 0)
            return false;

        else if (c == ')')
        {
            if (brackets.back() == '(')
                brackets.pop_back();
            else
                return false;
        }

        else if (c == '}')
        {
            if (brackets.back() == '{')
                brackets.pop_back();
            else
                return false;
        }

        else if (c == ']')
        {
            if (brackets.back() == '[')
                brackets.pop_back();
            else
                return false;
        }
    }

    return brackets.size() == 0;
    }
};

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();

        stack.add(s.charAt(0));

        for (int i = 1; i < s.length(); i++) {

            char c = s.charAt(i);

            if (c == 40 || c == 91 || c == 123)
                stack.add(c);
            else {

                if(stack.isEmpty())
                    return false;
                
                if (c == 41) {

                    if (stack.peek() == 40)
                        stack.pop();
                    else
                        return false;

                } else {
                    if (stack.peek() == c - 2)
                        stack.pop();
                    else
                        return false;
                }

            }

        }

        return stack.isEmpty();
        
    }
}